xbmc/pvr/windows/GUIWindowPVRGuide.cpp

```diff
@@ -97,11 +97,19 @@ void CGUIWindowPVRGuide::StopRefreshTimelineItemsThread()
 
 void CGUIWindowPVRGuide::RegisterObservers(void)
 {
+  CSingleLock lock(m_critSection);
   g_EpgContainer.RegisterObserver(this);
+  if (g_PVRTimers)
+    g_PVRTimers->RegisterObserver(this);
+  CGUIWindowPVRBase::RegisterObservers();
 }
 
 void CGUIWindowPVRGuide::UnregisterObservers(void)
 {
+  CSingleLock lock(m_critSection);
+  CGUIWindowPVRBase::UnregisterObservers();
+  if (g_PVRTimers)
```

Comment:
Your first statement is correct. If it does not access an invalid pointer, it may unregister at a different instance which is not correct either. I don't think we should switch to shared pointers as a solution, the problem is the observer pattern that is an anti pattern here.
I would use window messages, and maybe route them through a proxy object.